src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -300,13 +335,36 @@ class NBDServer
 
   void reader_entry()
   {
-    while (!terminated) {
+    struct pollfd poll_fds[2];
+    memset(poll_fds, 0, sizeof(struct pollfd) * 2);
+    poll_fds[0].fd = fd;
+    poll_fds[0].events = POLLIN;
+    poll_fds[1].fd = terminate_event_fd;
+    poll_fds[1].events = POLLIN;
+
+    while (true) {
       std::unique_ptr<IOContext> ctx(new IOContext());
       ctx->server = this;
 
       dout(20) << __func__ << ": waiting for nbd request" << dendl;
 
-      int r = safe_read_exact(fd, &ctx->request, sizeof(struct nbd_request));
+      int r = poll(poll_fds, 2, -1);
+      if (r < 0) {
+        derr << "failed to poll nbd: " << cpp_strerror(r) << dendl;
```

Comment:
Nit: ignore `errno == EINTR` errors and make sure you grab the error from `errno` instead of attempting to print `r`